packages/backend-openapi-utils/src/schema/parameter-validation.ts

```diff
@@ -0,0 +1,513 @@
+} from './types';
+import Ajv from 'ajv';
+import { OperationError, OperationParsingError } from './errors';
+import { mockttpToFetchRequest } from './utils';
+
+type ReferencelessSchemaObject = SchemaObject & { $ref?: never };
+
+type ReferencelessParameterObject = Omit<ParameterObject, 'schema'> & {
+  schema: ReferencelessSchemaObject;
+};
+
+class BaseParameterParser {
+  ajv: Ajv;
+  operation: Operation;
+  parameters: Record<string, ReferencelessParameterObject> = {};
+  constructor(
+    parameterIn: string,
+    operation: Operation,
+    options: ParserOptions,
+  ) {
+    this.ajv = options.ajv;
+    this.operation = operation;
+    const { schema, path, method } = operation;
+    const parameters = schema.parameters || [];
+    for (const parameter of parameters) {
+      if ('$ref' in parameter) {
+        throw new Error(
+          `[(${method}) ${path}] Reference objects are not supported`,
+        );
+      }
+
+      if (!parameter.schema) {
+        throw new OperationError(
+          operation,
+          'Schema not found for path parameter',
+        );
+      }
+      if ('$ref' in parameter.schema) {
+        throw new OperationError(
+          this.operation,
+          'Reference objects are not supported for parameters',
+        );
+      }
+      if (parameter.in === parameterIn) {
+        this.parameters[parameter.name] =
+          parameter as ReferencelessParameterObject;
+      }
+    }
+  }
+
+  /**
+   * Attempt to transform a string value to its expected type, this allows Ajv to perform validation. As these are parameters,
+   *  support for edge cases like nested type casting is not currently supported.
+   * @param value
+   * @param schema
+   * @returns
+   */
+  optimisticallyParseValue(value: string, schema: SchemaObject) {
+    if (schema.type === 'integer') {
+      return parseInt(value, 10);
+    }
+    if (schema.type === 'number') {
+      return parseFloat(value);
+    }
+    if (schema.type === 'boolean') {
+      if (['true', 'false'].includes(value)) {
+        return value === 'true';
+      }
+      throw new Error('Invalid boolean value must be either "true" or "false"');
+    }
+    return value;
+  }
+}
+
+const PLACE_A_BEFORE_B = -1;
+const PLACE_A_AFTER_B = 1;
+const EQUAL = 0;
+
+export class QueryParameterParser
+  extends BaseParameterParser
+  implements RequestParser<Record<string, any>>
+{
+  constructor(operation: Operation, options: ParserOptions) {
+    super('query', operation, options);
+  }
+  async parse(request: Request) {
+    const { searchParams } = new URL(request.url);
+    const remainingQueryParameters = new Set<string>(searchParams.keys());
+    const queryParameters: Record<string, any> = {};
+
+    const parameterIterator = Object.entries(this.parameters);
+
+    // object parameters with form/explode style should be processed last as they collect all remaining parameters.
+    parameterIterator.sort(([_, parameterA], [_B, parameterB]) => {
```

Comment:
Potential different approach here couple be to filter partition into two arrays and then concat, potentially quite a bit less code.

Reply:
nice! yeah that makes this way simpler, updated!